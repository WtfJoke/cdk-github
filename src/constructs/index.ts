export * from './action-environment-secret';
export * from './action-secret';
export * from './github-resource';
