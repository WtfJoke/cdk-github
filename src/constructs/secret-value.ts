// import { SecretsManagerSecretOptions } from 'aws-cdk-lib';


// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html#dynamic-references-considerations
// export class SecretValue {

//   /**
//      * Creates a `SecretValue` with a value which is dynamically loaded from AWS Secrets Manager.
//      * @param secretId The ID or ARN of the secret
//      * @param options Options
//      */
//   static fromSecretsManager(secretId: string, options?: SecretsManagerSecretOptions = {}): SecretValue {
//     if (options.versionStage && options.versionId) {
//       throw new Error(`verionStage: '${options.versionStage}' and versionId: '${options.versionId}' were both provided but only one is allowed`);
//     }

//     const parts = [
//       secretId,
//       'SecretString',
//       options.jsonField || '',
//       options.versionStage || '',
//       options.versionId || '',
//     ];
//     return new SecretValue();;
//   }
//   /**
//        * Use a secret value stored from a Systems Manager (SSM) parameter.
//        *
//        * @param parameterName The name of the parameter in the Systems Manager
//        * Parameter Store. The parameter name is case-sensitive.
//        *
//        * @param version An integer that specifies the version of the parameter to
//        * use. If you don't specify the exact version, AWS CloudFormation uses the
//        * latest version of the parameter.
//        */
//   static fromSsmSecure(parameterName: string, version?: string): SecretValue {
//     return new SecretValue();
//   }


//   constructor() {
//     console.log('hi');
//   }
// }
