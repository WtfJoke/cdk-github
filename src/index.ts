export * from './constructs';
export * from './types/exported';
