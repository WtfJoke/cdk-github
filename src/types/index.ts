export * from './action-environment-secret-event-props';
export * from './action-secret-event-props';
export * from './on-event-request';
