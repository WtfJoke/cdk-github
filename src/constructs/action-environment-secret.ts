import { CustomResource, Duration, Names, Stack } from 'aws-cdk-lib';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { ActionEnvironmentSecretHandlerFunction } from '../handler/secrets/action-environment-secrets';
import { ActionEnvironmentSecretEventProps } from '../types';
import { IGitHubRepository, ISecretString } from '../types/exported';

export interface ActionEnvironmentSecretProps {
  /**
   * The AWS secret in which the OAuth GitHub (personal) access token is stored.
   */
  readonly githubToken: ISecretString;

  /**
   * The GitHub repository information (owner and name).
   */
  readonly repository: IGitHubRepository;

  /**
   * This AWS secret value will be stored in GitHub as a secret (under the name of actionSecretName).
   */
  readonly secretToBeStored: ISecretString;

  /**
   * The GitHub secret name to be stored.
   */
  readonly actionSecretName: string;

  /**
   * The GithHub environment name which the secret should be stored in.
   */
  readonly environment: string;
}

export class ActionEnvironmentSecret extends Construct {
  constructor(scope: Construct, id: string, props: ActionEnvironmentSecretProps) {
    super(scope, id);
    const { githubToken, actionSecretName, repository, secretToBeStored, environment } = props;
    const stack = Stack.of(this);
    const awsRegion = stack.region;
    const shortId = Names.uniqueId(this).slice(-8);

    const handler = new ActionEnvironmentSecretHandlerFunction(this, 'CustomResourceHandler', {
      functionName: 'GitHubActionEnvironmentSecretCustomResourceHandler' + shortId,
      description: 'Handles the creation/deletion of a GitHub Action environment secret - created by cdk-github',
      architecture: Architecture.ARM_64,
      timeout: Duration.minutes(10),
    });

    githubToken.grantRead(handler);
    secretToBeStored.grantRead(handler);

    const provider = new Provider(this, 'CustomResourceProvider', {
      onEventHandler: handler,
      logRetention: RetentionDays.ONE_WEEK,
    });

    const githubRepositorySecretEventProps: ActionEnvironmentSecretEventProps = {
      environment,
      githubTokenSecret: githubToken.serialize(stack),
      repositoryOwner: repository.owner,
      repositoryName: repository.name,
      sourceSecret: secretToBeStored.serialize(stack),
      actionSecretName,
      awsRegion,
    };

    new CustomResource(this, 'CustomResource', {
      serviceToken: provider.serviceToken,
      resourceType: 'Custom::GitHubActionEnvironmentSecret',
      properties: githubRepositorySecretEventProps,
    });
  }
}
