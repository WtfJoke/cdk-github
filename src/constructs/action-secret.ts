import { CustomResource, Duration, Names, Stack } from 'aws-cdk-lib';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { ISecret } from 'aws-cdk-lib/aws-secretsmanager';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { ActionSecretHandlerFunction } from '../handler/secrets/action-secrets';
import { ActionSecretEventProps } from '../types';
import { IGitHubRepository } from '../types/exported';

export interface ActionSecretProps {
  /**
   * The AWS secret in which the OAuth GitHub (personal) access token is stored
   */
  readonly githubTokenSecret: ISecret;

  /**
   * The GitHub repository information (owner and name)
   */
  readonly repository: IGitHubRepository;

  /**
   * The GitHub secret name to be stored
   */
  readonly repositorySecretName: string;

  /**
   * This AWS secret value will be stored in GitHub as a secret (under the name of repositorySecretName)
   */
  readonly sourceSecret: ISecret;

  /**
   * The key of a JSON field to retrieve in sourceSecret.
   * This can only be used if the secret stores a JSON object.
   *
   * @default - returns all the content stored in the Secrets Manager secret.
   */
  readonly sourceSecretJsonField?: string;
}

export class ActionSecret extends Construct {
  constructor(scope: Construct, id: string, props: ActionSecretProps) {
    super(scope, id);
    const { githubTokenSecret, repositorySecretName, repository, sourceSecret, sourceSecretJsonField } = props;
    const awsRegion = Stack.of(this).region;
    const shortId = Names.uniqueId(this).slice(-8);

    const handler = new ActionSecretHandlerFunction(this, 'CustomResourceHandler', {
      functionName: 'GitHubActionSecretCustomResourceHandler' + shortId,
      description: 'Handles the creation/deletion of a GitHub Action (repository) secret - created by cdk-github',
      architecture: Architecture.ARM_64,
      timeout: Duration.minutes(10),
    });

    githubTokenSecret.grantRead(handler);
    sourceSecret.grantRead(handler);

    const provider = new Provider(this, 'SecretProvider', {
      onEventHandler: handler,
      logRetention: RetentionDays.ONE_WEEK,
    });

    const githubRepositorySecretEventProps: ActionSecretEventProps = {
      githubTokenSecret: githubTokenSecret.secretArn,
      repositoryOwner: repository.owner,
      repositoryName: repository.name,
      sourceSecretArn: sourceSecret.secretArn,
      sourceSecretJsonField,
      repositorySecretName,
      awsRegion,
    };

    new CustomResource(this, 'CustomResource', {
      serviceToken: provider.serviceToken,
      resourceType: 'Custom::GitHubActionSecret',
      properties: githubRepositorySecretEventProps,
    });
  }
}
