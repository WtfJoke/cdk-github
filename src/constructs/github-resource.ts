import { CustomResource, Duration, Names, Stack } from 'aws-cdk-lib';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { IParameter } from 'aws-cdk-lib/aws-ssm';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { GithubResourceHandlerFunction } from '../handler/github-resource';
import { GitHubResourceEventProps } from '../types';
import { ISecretString } from '../types/exported';

export interface GitHubResourceProps {
  /**
   * The AWS secret in which the OAuth GitHub (personal) access token is stored
   */
  readonly githubToken: ISecretString;


  /**
   * The GitHub api endpoint url for creating resources in format: `POST /repos/OWNER/REPO/issues`. This is called when the GitHubResource is created.
   *
   * Example:
   * ```
   * const createRequestEndpoint = 'POST /repos/octocat/Hello-World/issues'
   * ```
   */
  readonly createRequestEndpoint: string;

  /**
   * The GitHub api request payload for creating resources.
   * This is a JSON parseable string.
   *
   * Used for  @see {@link GitHubResourceProps#createRequestEndpoint}.
   *
   * Example:
   * ```
   * const createRequestPayload = JSON.stringify({ title: 'Found a bug', body: "I'm having a problem with this.", assignees: ['octocat'], milestone: 1, labels: ['bug'] })
   * ```
   */
  readonly createRequestPayload?: string;


  /**
   * Used to extract a value from the result of the createRequest(Endpoint) to be used in update/deleteRequests.
   *
   * Example: `"number"` (for the issue number)
   *
   * When this parameter is set and can be extracted from the result, the extracted value will be used for the PhyscialResourceId of the CustomResource.
   * Changing the parameter once the stack is deployed is not supported.
   */
  readonly createRequestResultParameter?: string;


  /**
   * The GitHub api endpoint url to update this resource in format: `POST /repos/OWNER/REPO/issues`. This is called when the GitHubResource is updated.
   *
   * In most of the cases you want to either omit this or use the same value as createRequestEndpoint.
   *
   * Example:
   * ```
   * const updateRequestEndpoint = 'PATCH repos/octocat/Hello-World/issues/1'
   * ```
   * If you want to use the  @see {@link GitHubResourceProps#createRequestResultParameter}, you can use the following syntax (assuming you have set createRequestResultParameter to `"number"`):
   * ```
   * const updateRequestEndpoint = 'PATCH repos/octocat/Hello-World/:number'
   * ```
   */
  readonly updateRequestEndpoint?: string;

  /**
    * The GitHub api request payload to update this resources.
    * This is a JSON parseable string.
    *
    * Used for  @see {@link GitHubResourceProps#createRequestEndpoint}.
    *
    * Example:
    * ```
    * const updateRequestPayload = JSON.stringify({ title: 'Found a bug', body: "I'm having a problem with this.", assignees: ['octocat'], milestone: 1, state: 'open', labels: ['bug'] })
    * ```
    */
  readonly updateRequestPayload?: string;


  /**
   * The GitHub api endpoint url to delete this resource in format: `POST /repos/OWNER/REPO/issues`. This is called when the GitHubResource is deleted/destroyed.
   *
   * Example:
   * ```
   * const deleteRequestEndpoint = 'PATCH repos/octocat/Hello-World/issues/1'
   * ```
   * If you want to use the  @see {@link GitHubResourceProps#createRequestResultParameter}, you can use the following syntax (assuming you have set createRequestResultParameter to `"number"`):
   * ```
   * const deleteRequestEndpoint = 'PATCH repos/octocat/Hello-World/:number'
   * ```
   */
  readonly deleteRequestEndpoint: string;

  /**
   * The GitHub api request payload to delete this resource.
   * This is a JSON parseable string.
   *
   * Used for  @see {@link GitHubResourceProps#deleteRequestEndpoint}.
   *
   * Example:
   * ```
   * const deleteRequestPayload = JSON.stringify({ state: 'closed' })
   * ```
   */
  readonly deleteRequestPayload?: string;

  /**
   * The response body of the last GitHub api request will be written to this ssm parameter.
   */
  readonly writeResponseToSSMParameter?: IParameter;
}

export class GitHubResource extends Construct {
  constructor(scope: Construct, id: string, props: GitHubResourceProps) {
    super(scope, id);
    const {
      createRequestEndpoint, createRequestPayload,
      updateRequestEndpoint, updateRequestPayload,
      deleteRequestEndpoint, deleteRequestPayload,
      githubToken, createRequestResultParameter,
      writeResponseToSSMParameter,
    } = props;
    const stack = Stack.of(this);
    const awsRegion = stack.region;
    const shortId = Names.uniqueId(this).slice(-8);

    const handler = new GithubResourceHandlerFunction(this, 'CustomResourceHandler', {
      functionName: 'GitHubResourceCustomResourceHandler' + shortId,
      description: 'Handles the creation/deletion of a custom GitHub Resource - created by cdk-github',
      architecture: Architecture.ARM_64,
      timeout: Duration.minutes(10),
    });

    githubToken.grantRead(handler);
    writeResponseToSSMParameter?.grantWrite(handler);

    const provider = new Provider(this, 'Provider', {
      onEventHandler: handler,
      logRetention: RetentionDays.ONE_WEEK,
    });

    const githubResourceEventProps: GitHubResourceEventProps = {
      githubTokenSecret: githubToken.serialize(stack),
      createRequestEndpoint,
      createRequestPayload,
      createRequestResultParameter,
      updateRequestEndpoint,
      updateRequestPayload,
      deleteRequestEndpoint,
      deleteRequestPayload,
      responseBodySSMParameterName: writeResponseToSSMParameter?.parameterName,
      awsRegion,
    };

    new CustomResource(this, 'CustomResource', {
      serviceToken: provider.serviceToken,
      resourceType: 'Custom::GitHubResource',
      properties: githubResourceEventProps,
    });
  }
}
