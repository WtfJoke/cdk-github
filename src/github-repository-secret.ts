import * as path from 'path';
import { CustomResource, Duration, Stack } from 'aws-cdk-lib';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { ISecret } from 'aws-cdk-lib/aws-secretsmanager';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

export interface GitHubRepositorySecretProps {
  /**
   * The AWS secret in which the OAuth GitHub (personal) access token is stored
   */
  readonly githubTokenSecret: ISecret;

  /**
   * The GitHub repository owner
   */
  readonly repositoryOwner: string;

  /**
   * The GitHub repository name
   */
  readonly repositoryName: string;

  /**
   * The GitHub secret name to be stored
   */
  readonly repositorySecretName: string;

  /**
   * The AWS secret which should be stored as a GitHub as a secret
   */
  readonly sourceSecret: ISecret;
}

export class GitHubRepositorySecret extends Construct {
  constructor(scope: Construct, id: string, props: GitHubRepositorySecretProps) {
    super(scope, id);
    const { githubTokenSecret, repositorySecretName, repositoryName, repositoryOwner, sourceSecret } = props;
    const awsRegion = Stack.of(this).region;

    const handler = new NodejsFunction(this, 'CustomResourceHandler', {
      functionName: 'GitHubRepositorySecretCustomResourceHandler',
      description: 'Handles the creation/deletion of a GitHub repository secret - created by cdk-github',
      runtime: Runtime.NODEJS_16_X,
      entry: path.join(__dirname, 'handler', 'github-repository-secret', 'github-repository-secret-handler.ts'),
      architecture: Architecture.ARM_64,
      timeout: Duration.minutes(10),
    });

    githubTokenSecret.grantRead(handler);
    sourceSecret.grantRead(handler);

    const provider = new Provider(this, 'SecretProvider', {
      onEventHandler: handler,
      logRetention: RetentionDays.ONE_WEEK,
    });

    new CustomResource(this, 'CustomResource', {
      serviceToken: provider.serviceToken,
      resourceType: 'Custom::GithubRepositorySecret',
      properties: {
        githubTokenSecret: githubTokenSecret.secretArn,
        repositoryOwner,
        repositoryName,
        sourceSecretArn: sourceSecret.secretArn,
        repositorySecretName,
        awsRegion,
      },
    });
  }
}
