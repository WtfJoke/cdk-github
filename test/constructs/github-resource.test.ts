import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import { GitHubResource } from '../../src';
import { GitHubResourceEventProps } from '../../src/types';

describe('GithubResourceTestStack', () => {
  let template: Template;

  beforeAll(() => {
    template = getTemplate();
  });

  it('Should include custom resource handler', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      FunctionName: Match.stringLikeRegexp('GitHubResourceCustomResourceHandler'),
      Description: Match.stringLikeRegexp('created by cdk-github'),
    });
  });

  it('Should include ssm parameter', () => {
    template.hasResourceProperties('AWS::SSM::Parameter', {
      Name: '/cdk-github/github-resource/GitHubResourceIssueTestStack',
      Type: 'String',
      Value: 'UNFILLED',
    });
  });

  it('Should include github resource', () => {
    template.hasResourceProperties('Custom::GitHubResource', {
      createRequestEndpoint: 'POST /repos/WtfJoke/dummytest/issues',
      createRequestPayload: defaultProps.createRequestPayload,
      createRequestResultParameter: 'number',
      deleteRequestEndpoint: 'PATCH /repos/WtfJoke/dummytest/issues/:number',
      deleteRequestPayload: '{"state":"closed"}',
      createRequestResultValueSSMParameterName: Match.anyValue(),
    });
  });

  it('Should include github resource with update endpoint', () => {
    const updateProps = { ...defaultProps, updateRequestEndpoint: 'PATCH /repos/WtfJoke/dummytest/issues/:number', updateRequestPayload: JSON.stringify({ body: "I'm editing this issue ;)" }) };
    getTemplate(updateProps).hasResourceProperties('Custom::GitHubResource', {
      createRequestEndpoint: 'POST /repos/WtfJoke/dummytest/issues',
      createRequestPayload: defaultProps.createRequestPayload,
      createRequestResultParameter: 'number',
      updateRequestEndpoint: 'PATCH /repos/WtfJoke/dummytest/issues/:number',
      updateRequestPayload: '{"body":"I\'m editing this issue ;)"}',
      deleteRequestEndpoint: 'PATCH /repos/WtfJoke/dummytest/issues/:number',
      deleteRequestPayload: '{"state":"closed"}',
      createRequestResultValueSSMParameterName: Match.anyValue(),
    });
  });

  const getTemplate = (props?: GithubResourceTestStackProps) => {
    const app = new App();
    const stack = new GithubResourceTestStack(app, 'GitHubResourceIssueTestStack', props ? props : defaultProps);

    return Template.fromStack(stack);
  };
});

type GithubResourceTestStackProps = StackProps & Partial<GitHubResourceEventProps>

const defaultProps: GitHubResourceEventProps = {
  githubTokenSecret: 'GITHUB_TOKEN',
  awsRegion: 'eu-central-1',
  createRequestEndpoint: 'POST /repos/WtfJoke/dummytest/issues',
  createRequestPayload: JSON.stringify({ title: 'Testing cdk-github', body: "I'm opening an issue by using aws cdk ðŸŽ‰", labels: ['bug'] }),
  createRequestResultParameter: 'number',
  deleteRequestEndpoint: 'PATCH /repos/WtfJoke/dummytest/issues/:number',
  deleteRequestPayload: JSON.stringify({ state: 'closed' }),
};

class GithubResourceTestStack extends Stack {
  constructor(scope: Construct, id: string, props: GithubResourceTestStackProps) {
    super(scope, id, props);
    const {
      githubTokenSecret = defaultProps.githubTokenSecret,
      createRequestEndpoint = defaultProps.createRequestEndpoint,
      createRequestPayload = defaultProps.createRequestPayload,
      createRequestResultParameter = defaultProps.createRequestResultParameter,
      createRequestResultValueSSMParameterName = defaultProps.createRequestResultValueSSMParameterName,
      deleteRequestEndpoint = defaultProps.deleteRequestEndpoint,
      deleteRequestPayload = defaultProps.deleteRequestPayload,
      updateRequestEndpoint,
      updateRequestPayload,
    } = props;

    new GitHubResource(this, 'Issue', {
      createRequestEndpoint,
      createRequestPayload,
      createRequestResultParameter,
      deleteRequestEndpoint,
      deleteRequestPayload,
      updateRequestEndpoint,
      updateRequestPayload,
      writeResponseToSSMParameter: createRequestResultValueSSMParameterName ? StringParameter.fromStringParameterName(this, 'resultParam', createRequestResultValueSSMParameterName) : undefined,
      githubTokenSecret: Secret.fromSecretNameV2(this, 'ghSecret', githubTokenSecret),
    });
  }
}
