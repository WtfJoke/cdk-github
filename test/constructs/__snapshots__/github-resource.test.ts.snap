// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubResourceTestStack Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "IssueCustomResourceCB0851C4": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "IssueProviderframeworkonEventDB50811A",
            "Arn",
          ],
        },
        "awsRegion": Object {
          "Ref": "AWS::Region",
        },
        "createRequestEndpoint": "POST /repos/WtfJoke/dummytest/issues",
        "createRequestPayload": "{\\"title\\":\\"Testing cdk-github\\",\\"body\\":\\"I'm opening an issue by using aws cdk ðŸŽ‰\\",\\"labels\\":[\\"bug\\"]}",
        "createRequestResultParameter": "number",
        "deleteRequestEndpoint": "PATCH /repos/WtfJoke/dummytest/issues/:number",
        "deleteRequestPayload": "{\\"state\\":\\"closed\\"}",
        "githubTokenSecret": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":secret:GITHUB_TOKEN",
            ],
          ],
        },
      },
      "Type": "Custom::GitHubResource",
      "UpdateReplacePolicy": "Delete",
    },
    "IssueCustomResourceHandler9A5608E1": Object {
      "DependsOn": Array [
        "IssueCustomResourceHandlerServiceRoleDefaultPolicyFC76DDEF",
        "IssueCustomResourceHandlerServiceRoleDC4FD92B",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c0628540c4623b503f12bb232b5194b250f6353c8a8571704950a6ee3f2662e7.zip",
        },
        "Description": "Handles the creation/deletion of a custom GitHub Resource - created by cdk-github",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "GitHubResourceCustomResourceHandlerF8A6DFD5",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IssueCustomResourceHandlerServiceRoleDC4FD92B",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 600,
      },
      "Type": "AWS::Lambda::Function",
    },
    "IssueCustomResourceHandlerServiceRoleDC4FD92B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "IssueCustomResourceHandlerServiceRoleDefaultPolicyFC76DDEF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:GITHUB_TOKEN-??????",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "IssueCustomResourceHandlerServiceRoleDefaultPolicyFC76DDEF",
        "Roles": Array [
          Object {
            "Ref": "IssueCustomResourceHandlerServiceRoleDC4FD92B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "IssueProviderframeworkonEventDB50811A": Object {
      "DependsOn": Array [
        "IssueProviderframeworkonEventServiceRoleDefaultPolicyF89FE6D1",
        "IssueProviderframeworkonEventServiceRoleD8E0B8D0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eaeb69bc290b516fe3b049f89d6118b22249df682fbabf56af300cf345198574.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (GitHubResourceIssueTestStack/Issue/Provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "IssueCustomResourceHandler9A5608E1",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "IssueProviderframeworkonEventServiceRoleD8E0B8D0",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "IssueProviderframeworkonEventLogRetention988B5F00": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "IssueProviderframeworkonEventDB50811A",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "IssueProviderframeworkonEventServiceRoleD8E0B8D0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "IssueProviderframeworkonEventServiceRoleDefaultPolicyF89FE6D1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "IssueCustomResourceHandler9A5608E1",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "IssueCustomResourceHandler9A5608E1",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "IssueProviderframeworkonEventServiceRoleDefaultPolicyF89FE6D1",
        "Roles": Array [
          Object {
            "Ref": "IssueProviderframeworkonEventServiceRoleD8E0B8D0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e5032a7ed03a7762b99560439c17f58d046a35d40ebf64ba70c63157df388563.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
